go/parquet/internal/utils/bit_reader.go

```diff
@@ -0,0 +1,344 @@
+// Returning false if there aren't enough bytes remaining in the stream or if an invalid
+// type is passed. The bytes are read aligned to byte boundaries.
+//
+// v must be a pointer to a byte or sized uint type (*byte, *uint16, *uint32, *uint64).
+// encoded values are assumed to be little endian.
+func (b *BitReader) GetAligned(nbytes int, v interface{}) bool {
+	// figure out the number of bytes to represent v
+	typBytes := int(reflect.TypeOf(v).Elem().Size())
+	if nbytes > typBytes {
+		return false
+	}
+
+	bread := bitutil.BytesForBits(int64(b.bitoffset))
+
+	b.byteoffset += bread
+	n, err := b.reader.ReadAt(b.raw[:nbytes], b.byteoffset)
+	if err != nil && err != io.EOF {
+		return false
+	}
+	if n != nbytes {
+		return false
+	}
+	// zero pad the the bytes
+	memory.Set(b.raw[n:typBytes], 0)
+
+	switch v := v.(type) {
+	case *byte:
+		*v = b.raw[0]
+	case *uint64:
+		*v = binary.LittleEndian.Uint64(b.raw[:typBytes])
+	case *uint32:
+		*v = binary.LittleEndian.Uint32(b.raw[:typBytes])
+	case *uint16:
+		*v = binary.LittleEndian.Uint16(b.raw[:typBytes])
+	default:
+		return false
+	}
+
+	b.byteoffset += int64(nbytes)
+
+	b.bitoffset = 0
+	b.fillbuffer()
+	return true
+}
+
+// fillbuffer fills the uint64 buffer with bytes from the underlying stream
+func (b *BitReader) fillbuffer() error {
+	n, err := b.reader.ReadAt(b.raw[:], b.byteoffset)
+	if err != nil && n == 0 && err != io.EOF {
+		return err
+	}
+	for i := n; i < 8; i++ {
+		b.raw[i] = 0
+	}
+	b.buffer = binary.LittleEndian.Uint64(b.raw[:])
+	return nil
+}
+
+// next reads an integral value from the next bits in the buffer
+func (b *BitReader) next(bits uint) (v uint64, err error) {
+	v = trailingBits(b.buffer, b.bitoffset+bits) >> b.bitoffset
+	b.bitoffset += bits
+	// if we need more bits to get what was requested then refill the buffer
+	if b.bitoffset >= 64 {
+		b.byteoffset += 8
+		b.bitoffset -= 64
+		if err = b.fillbuffer(); err != nil {
+			return 0, err
+		}
+		v |= trailingBits(b.buffer, b.bitoffset) << (bits - b.bitoffset)
+	}
+	return
+}
+
+// GetBatchIndex is like GetBatch but for IndexType (used for dictionary decoding)
+func (b *BitReader) GetBatchIndex(bits uint, out []IndexType) (i int, err error) {
+	if bits > 32 {
+		return 0, errors.New("must be 32 bits or less per read")
+	}
+
+	var val uint64
+
+	length := len(out)
+	// if we're not currently byte-aligned, read bits until we are byte-aligned.
+	for ; i < length && b.bitoffset != 0; i++ {
+		val, err = b.next(bits)
+		out[i] = IndexType(val)
+		if err != nil {
+			return
+		}
+	}
+
+	b.reader.Seek(b.byteoffset, io.SeekStart)
+	// grab as many 32 byte chunks as possible in one shot
+	if i < length { // IndexType should be a 32 bit value so we can do quick unpacking right into the output
+		numUnpacked := unpack32(b.reader, (*(*[]uint32)(unsafe.Pointer(&out)))[i:], int(bits))
+		i += numUnpacked
+		b.byteoffset += int64(numUnpacked * int(bits) / 8)
```

Comment:
bits should never be large enough where it matters whether or not it's 32/64 bit so using the default `int` is safe and fine here. since `numUnpacked` is an int, if we did `int32(bits)` we'd have to also do `int32(numUnpacked)` so i opted for what looked cleanest to me rather than doing something like `int64(int32(numUnpacked) * int32(bits) / 8)` or `int64(numUnpacked) * int64(bits) / 8`.

Reply:
ok